keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,226 @@
+
+We will extend ClusterQueueSpec definition by adding a reference to
+AdmissionChecks that the queue needs to perform before admitting a
+workload. The design of AdmissionChecks is similar to IngressClass
+- it allows passing additional parameters, specific to particular checks
+that need to be performed. 
+
+```
+// ClusterQueueSpec defines the desired state of ClusterQueue
+type ClusterQueueSpec struct {
+[...]
+	// List of AdmissionCheck names that needs to be passed
+	// to admit a workload.  
+	AdmissionChecks []string `json:"addotomalChecks"`
+}
+
+
+// Cluster-scoped top-level object defining a check that 
+// can be referenced in ClusterQueue
+type AdmissionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec AdmissionCheckSpec `json:"spec,omitempty"`
+}
+
+const (
+	Retry CheckFailureState = “Retry”
+	Reject CheckFailureState = “Reject”
+)
+
+type AdmissionCheckSpec struct {
+// Name of the controller which will actually perform
+// the checks. Cannot be empty.
+	ControllerName string `json:"controllerName"`
+
+	// How long to keep the workload suspended 
+	// after a failed check.
+	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`
+
+	// A reference to the additional parameters for the check. 
+	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
+}
+
+// Points to where the parameters for the checks are. 
+// As clusterqueue are in cluster scope - this should be
+// a dedicated CRD specific for the Controller.
+type AdmissionCheckParametersReference struct {
+	// ApiGroup is the group for the resource being referenced. 
+	APIGroup string `json:"apiGroup"`
+
+	// Kind is the type of the resource being referenced.
+	Kind string `json:"kind""`
+
+	// Name is the name of the resource being referenced.
+	Name string `json:"name"`
+}
+```
+
+For every workload that is put to a ClusterQueue, Kueue will perform the very 
+same checks that it does today, before admitting a workload. However, once the basic checks pass AND there are some AdmissionChecks configured, AND the workload is not in on-hold retry state from some check, it will:
+
+1. Fill the Admission field in workload, with the desired flavor assignment. 
+2. Not do any preemptions yet.
+3. Set conditions on the workload:
+* “AdmissionPrecheck” to true
+* “AdmissionCheck|<checkName>” to unknown for each of the add
+
+Kueue will only pass as many pods past AdmissionPrecheck as there would
+fit in the quota together, possibly limited by a Kueue configuration option.
+That would violate a bit BestEffort logic of queues - if there is 1000 tasks
+in a queue that doesn’t pass quota and 1001st task passes, the best effort
+queue would let it in. BestEffort queue with 1000 tasks that don’t pass
+AdditioanlChecks would not let 1001st task in (it will not reach the
+checks). Without this limitation, a large number of tasks could be switched
+back and forth from Precheck to suspended state. 
```

Comment:
When we have determined that checks for a workload don't pass (and the retry period is over), we should probably put the workload in the `inadmissibleWorkloads` set.
This way, BestEffort would mostly work the same way as before, except that it's, potentially, slower.

Reply:
Ack.